### ISSUE TITLE
Optional nested records

### ISSUE THREAD

**Issue by tricon** [aconchillo/guile-json#78]
Consider the following:

```scheme
(define-json-type <account>
  (id)
  (username)
  (omitted "omitted" <omitted>)
  (boolean))

(define-json-type <omitted>
  (name))


(define test-json-account
  "{\"id\":\"11111\",\"username\":\"jane\",\"boolean\":false}")

(define test-account (json->account test-json-account))
```

Executing this returns:

```
In unknown file:
           0 (assoc "name" #<unspecified>)

ERROR: In procedure assoc:
In procedure assoc: Wrong type argument in position 2 (expecting association list): #<unspecified>
```

Could we get support for records in which a nested record may be missing from incoming JSON?

P.S. This library is really well designed and is a joy to use. Thank you for your work!

**Comment by aconchillo** [comment]
Hi @tricon ! Thank you for the report and sorry for the long delay, many things going on lately. I think I'll be able to take a look some time this week. Hang in there!

**Comment by aconchillo** [comment]
This should now be fixed upstream. I'll release a version soon.

**Comment by tricon** [comment]
You, good sir, are a gentleman and a scholar. Thank you for your work and this wonderful module.

Apr 19, 2022 00:03:11 Aleix Conchillo Flaqué ***@***.***>:

**Comment by aconchillo** [comment]
> You, good sir, are a gentleman and a scholar. Thank you for your work and this wonderful module.

I wish all that was true :smile:. For now, I just released version 4.7.1 which only includes this fix. Thanks again for filing the issue.

**Comment by emixa-d** [comment]
I think that commit broke my code after updating guile-json to 4.7.1.  I've a `define-json-mapping`:

```
(define (or-constant constant)
  (lambda (proc)
    (lambda (foo)
      (if (unspecified? foo)
	  constant
	  (proc foo)))))
(define or-emptystring* ((or-constant "") identity))
(define-json-mapping <package> make-package package?
  %json->package <=> %package->json <=> scm->package <=> package->scm
  [...]
  (license-file package-license-file "license-file" or-emptystring*) ; string
  [...])
```

The idea is that the field 'license-file' is set to "" if not present in the JSON.

Likewise, I've a record <manifest> that has a record field 'lib' that is set to `#false` when not present in the JSON (instead of the not very Schemy *unspecified*, which cannot directly be used in `and=>`, '`if` and `or`):
 
```
(define-json-mapping <manifest> make-manifest manifest?
  %json->manifest <=> %manifest->json <=> scm->manifest <=> manifest->scm
  (package manifest-package "package" scm->package)
  (lib manifest-lib "lib" (or-false scm->target))
```

So it seems to me that optional nested records were already implemented, albeit not with a very convenient or documented interface, and this was an incompatible change, even though only the minor version was changed and not the major version.

**Comment by emixa-d** [comment]
To be clear, for now I've worked around the incompatible change by pinning the old version, and it should be feasible to change the code to the new guile-json version. However, being able to have (non-unspecified) default values seems neat to me.

**Comment by aconchillo** [comment]
@emixa-d oops, you are totally right, 4.7.1 is not backwards compatible. I didn't realize that when I made the fix. So I'm reopening this issue.

**Comment by aconchillo** [comment]
So, I rolled back the breaking change and fixed the original issue properly.

The problem was that since `omitted` field is `*unspecified*` we don't want to call `(scm->omitted *unspecified*)`. This only happens in `(define-json-type)` where we can specify a sub-field type.

**Comment by aconchillo** [comment]
Just released 4.7.2.

@emixa-d please, let me know if this fixes your issue.
